noteable_magics/sql/meta_commands.py

```diff
@@ -483,6 +501,353 @@ def _split_schema_table(schema_table: str) -> Tuple[Optional[str], str]:
+
+        If no primary key index is defined, will return None for name, empty list for columns.
+        """
+        primary_index_dict = inspector.get_pk_constraint(relation_name, schema_name)
+
+        pkey_name = primary_index_dict.get('name', '(unnamed primary key)')
+
+        if primary_index_dict['constrained_columns']:
+            return pkey_name, primary_index_dict['constrained_columns']
+        else:
+            return '', []  # None, []
+
+    def introspect_columns(
+        self, inspector: 'SchemaStrippingInspector', schema_name: str, relation_name: str
+    ) -> List[ColumnModel]:
+
+        column_dicts = inspector.get_columns(relation_name, schema=schema_name)
+
+        retlist = []
+
+        for col in column_dicts:
+            comment = col.get('comment')  # Some dialects do not return.
+
+            try:
+                type_name = str(col['type'].as_generic()).lower()
+            except NotImplementedError:
+                # ENG-5268: More esoteric types like UUID do not implement .as_generic()
+                type_name = str(col['type']).replace('()', '').lower()
+
+            retlist.append(
+                ColumnModel(
+                    name=col['name'],
+                    is_nullable=col['nullable'],
+                    data_type=type_name,
+                    default_expression=col['default'],
+                    comment=comment,
+                )
+            )
+
+        return retlist
+
+    def inform_gate_start(self, auth_header: dict, datasource_id: UUID):
+        """Tell gate to forget about any prior structures known for this datasource"""
+
+        # No route implemented for this yet, but need one, otherwise Gate-side will
+        # never forget about dropped tables/views.
+
+        pass
+
+    def inform_gate_relation(
+        self,
+        auth_header: dict,
+        datasource_id,
+        relation_description: RelationStructureDescription,
+    ):
+        """POST this `relation_description` up to Gate for storage, relating to `datasource_id`"""
+
+        as_dict_from_json = json.loads(relation_description.json())
+
+        resp = requests.post(
+            f"http://gate.default/api/v1/datasources/{datasource_id}/schema/relation",
+            json=as_dict_from_json,
+            headers=auth_header,
+        )
+
+        if resp.status_code == 204:
+            print(
+                f'Stored structure of {relation_description.schema_name}.{relation_description.relation_name}'
+            )
+        else:
+            print(
+                f'Failed storing structure of {relation_description.schema_name}.{relation_description.relation_name}: {resp.status_code}, {resp.text}'
+            )
+
+    def get_datasource_id(self) -> UUID:
+        """Convert a noteable_magics.sql.connection.Connection's name to the original
+        UUID Gate knew it as.
+
+        Will fail with ValueError if attempted against a legacy datasource handle like '@noteable', so
```

Comment:
What about raising a more human readable error for the user to see if they do try to introspect something else?

Maybe when trying to introspect `@noteable` or `@foo`: `f"Invalid datasource id {handle[1:]!r}"`

Reply:
Gonna check real early, then bail with 'Cannot introspect into legacy resource.' if not coercable to a uuid.